### ISSUE TITLE
Figure our `lock` option in rioxarray.open_rasterio

### ISSUE THREAD

**Issue by blindjesse** [csp-inc/lumonitor#20]
I'm trying to optimize performance of cog reading in an IterableDataset. In particular, I'm trying to read samples from a list of cog files using multiple workers. Multiple workers may be trying to read the same file at the same time. When using `rioxarray.open_rasterio` with multiple workers and chunking, but no lock set, this results in an error. 

Consider https://github.com/csp-inc/lumonitor/blob/dev/src/model/StreamingDataset.py, in particular the `riox.open_rasterio` options in this function:

```
def read_tile(path, chunks):
    return rioxr.open_rasterio(path, chunks=chunks).fillna(0)
```
First if chunks are not set, it caches the whole file in memory, which isn't going to work when opening many files at all (try it and see). If I leave it like this and create an instance with 1 worker, e.g (from https://github.com/csp-inc/lumonitor/blob/dev/src/model/streaming_model.py):

```
szd = StreamingDataset(
    training_files,
    label_files,
    label_band=1,
    feature_chip_size=512,
    label_chip_size=70,
    num_chips_per_tile=200
)

loader = DataLoader(
    szd,
    num_workers=1,
    batch_size=18,
    pin_memory=True
)
```
Then things go fine. But if I increase the number of workers, I get something like

```
ERROR 1: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~93
ERROR 1: TIFFFillTile:Read error at row 4294967295, col 4294967295, tile 93; got 2591 bytes, expected 2789307
ERROR 1: TIFFReadEncodedTile() failed.
ERROR 1: data/cog/2016/training/10TDK.tif, band 1: IReadBlock failed at X offset 3, Y offset 6: TIFFReadEncodedTile() failed.
ERROR 1: LZWDecode:Wrong length of decoded string: data probably corrupted at scanline 768
ERROR 1: TIFFReadEncodedTile() failed.
ERROR 1: data/cog/2016/training/10TDK.tif, band 1: IReadBlock failed at X offset 3, Y offset 6: TIFFReadEncodedTile() failed.
ERROR 1: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~93
ERROR 1: TIFFFillTile:Read error at row 4294967295, col 4294967295, tile 93; got 1047071 bytes, expected 2789307
ERROR 1: TIFFReadEncodedTile() failed.
ERROR 1: TIFFFillTile:Read error at row 4294967295, col 4294967295, tile 93; got 0 bytes, expected 1299291031
ERROR 1: TIFFReadEncodedTile() failed.
```
etc.

I believe this is because Gdal cannot have multiple reads of the same tiff at the same time (from google search, but also see e.g. https://corteva.github.io/rioxarray/stable/examples/read-locks.html). 

After reading [this issue](https://github.com/corteva/rioxarray/issues/214), the docs (see above), and the code, it looks like I need to set `lock=False` or `lock=<a lock>` (docs say that `lock=True` will work too, but see https://github.com/corteva/rioxarray/issues/273). If I set `lock=False`, it works but it's very slow, I think because the cog file is being re-opened on each tile pull (as discussed in those issues).

So I want to set  the lock to something that will allow the file to only be opened once, but dis-allow concurrent reads of the same file (so I guess the workers need to keep track of whether another worker is reading that file and wait/move on????).

**Comment by blindjesse** [comment]
One idea which would ~sidestep this issue: https://stackoverflow.com/questions/57729279/how-to-ensure-that-a-batch-contains-samples-from-all-workers-with-pytorchs-data

**Comment by blindjesse** [comment]
Performance is adequate using rasterio.open and using `shuffle=True` on dataloader shuffle file paths
